File: src/main/java/de/tlmrgvf/drtd/utils/bch/FiniteField.java
/*
 *
 * BSD 2-Clause License
 *
 * Copyright (c) 2020, Till Mayer
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package de.tlmrgvf.drtd.utils.bch;

import de.tlmrgvf.drtd.Drtd;

import java.util.logging.Logger;

public final class FiniteField {
    private final static Logger LOGGER = Drtd.getLogger(FiniteField.class);
    private final long[] primitiveRoots;
    private final int[] primitiveRootsLog;
    private final int elements;
    private final int exponent;

    public FiniteField(int exponent) {
        this.exponent = exponent;

        //Or else we won't be able to use an array to map roots to their exponents
        assert exponent <= Integer.SIZE - 1;
        assert exponent > 0;

        LOGGER.finer("Starting search for an irreducible polynomial of degree " + exponent + "...");
        int irreducibleSearch = 1 << exponent;
        Z2Polynomial irreducible = null;

        /*
         * REALLY primitive way of searching for an irreducible polynomial, basically just like the primitive prime
         * number check, but the exponent should not get too big with BCH codes and this is only calculated once
         */
        for (int i = 0; i < irreducibleSearch; ++i) {
            final Z2Polynomial irreducibleCandidate = new Z2Polynomial(irreducibleSearch ^ i);

            boolean isIrreducible = true;
            /* Start at x^1 as every polynomial is divisible by 1 */
            for (int test = 2; test < (irreducibleSearch ^ (irreducibleSearch - 1)); ++test) {
                Z2Polynomial remainder = irreducibleCandidate.remainder(new Z2Polynomial(test));
                if (remainder.getCoefficients() != 0 && test == irreducibleCandidate.getCoefficients() ||
                        remainder.getCoefficients() == 0 && test != irreducibleCandidate.getCoefficients()) {
                    isIrreducible = false;
                    break;
                }
            }

            if (isIrreducible) {
                LOGGER.finer("Found irreducible polynomial: " + irreducibleCandidate);
                irreducible = irreducibleCandidate;
                break;
            }
        }

        /* This should never happen, I think? */
        if (irreducible == null)
            throw new IllegalStateException("Could not find an irreducible!");

        primitiveRoots = new long[(int) (Math.pow(2, exponent) - 1)];
        elements = primitiveRoots.length;
        primitiveRootsLog = new int[primitiveRoots.length + 1];

        long alpha = 1; //Use x as the primitive root of unity
        primitiveRoots[0] = alpha; //a^0
        primitiveRootsLog[0] = -1; //There is no exponent to make a^n = 0!

        for (int i = 1; i < primitiveRoots.length; ++i) {
            alpha <<= 1; //times x
            primitiveRoots[i] = new Z2Polynomial(alpha).remainder(irreducible).getCoefficients();
            primitiveRootsLog[(int) primitiveRoots[i]] = i;
        }
    }

    public long[] getPrimitiveRoots() {
        return primitiveRoots;
    }

    public int getPrimitiveRootCount() {
        return elements;
    }

    public int getRootExponent(Z2Polynomial root) {
        assert root.degree() < exponent;
        return primitiveRootsLog[(int) root.getCoefficients()];
    }

    public Z2Polynomial multiplyRoots(Z2Polynomial multiplier, Z2Polynomial multiplicand) {
        assert multiplier.degree() < exponent && multiplicand.degree() < exponent;
        if (multiplier.isZero() || multiplicand.isZero()) return new Z2Polynomial(0);

        int alphaExponent = primitiveRootsLog[(int) multiplier.getCoefficients()]
                + primitiveRootsLog[(int) multiplicand.getCoefficients()];
        return new Z2Polynomial(primitiveRoots[alphaExponent % primitiveRoots.length]);
    }

    public Z2Polynomial powerOfRoot(Z2Polynomial root, int exponent) {
        assert root.degree() < this.exponent;

        int alphaExponent = primitiveRootsLog[(int) root.getCoefficients()];
        return new Z2Polynomial(primitiveRoots[(alphaExponent * exponent) % primitiveRoots.length]);
    }

    public Z2Polynomial powerOfX(int exponent) {
        return new Z2Polynomial(primitiveRoots[exponent % primitiveRoots.length]);
    }

    public Z2Polynomial rootInverse(Z2Polynomial root) {
        assert !root.isZero(); //There is no multiplicative inverse to 0

        int alphaExponent = primitiveRootsLog[(int) root.getCoefficients()];
        return new Z2Polynomial(primitiveRoots[(elements - alphaExponent) % primitiveRoots.length]);
    }

    public Z2Polynomial calculateSyndrome(Z2Polynomial polynomial, int n) {
        Z2Polynomial result = new Z2Polynomial();
        if (polynomial.getExponentValues() == null)
            return null;

        for (int exponent : polynomial.getExponentValues()) result = result.add(powerOfX(exponent * n));
        return result;
    }
}

File: src/util/bch/FiniteField.cpp
#include "FiniteField.hpp"

using namespace Bch;

FiniteField::FiniteField(u8 exponent)
    : m_exponent(exponent) {
    assert(exponent < Z2Polynomial::coefficient_count);
    s_log.info() << "Starting search for an irreducible polynomial of degree " << static_cast<int>(exponent) << "...";
    Z2Polynomial::Store search = 1 << exponent;
    std::optional<Z2Polynomial> irreducible;

    /*
     * REALLY primitive way of searching for an irreducible polynomial, basically just like the primitive prime
     * number check, but the exponent should not get too big with BCH codes and this is only calculated once
     */
    for (Z2Polynomial::Store i = 0; i < search; ++i) {
        const Z2Polynomial candidate(search ^ i);
        bool is_irreducible = true;

        /* Start at x^1 as every polynomial is divisible by 1 */
        for (Z2Polynomial::Store test = 2; test < (search ^ (search - 1)); ++test) {
            const auto remainder = candidate % test;
            if ((!remainder.is_zero() && test == candidate.coefficients()) || (remainder.is_zero() && test != candidate.coefficients())) {
                is_irreducible = false;
                break;
            }
        }

        if (is_irreducible) {
            s_log.info() << "Found irreducible polynomial 0x" << std::hex << candidate.coefficients();
            irreducible = candidate;
            break;
        }
    }

    /* This should never fail, I think? */
    assert(irreducible.has_value());

    m_element_count = Util::pow2(exponent) - 1;
    m_primitive_roots = Buffer<Z2Polynomial>(m_element_count);
    m_primitive_roots_log = Buffer<size_t>(m_element_count + 1);

    Z2Polynomial::Store alpha = 1; //Use x as the primitive root of unity
    m_primitive_roots[0] = alpha;  //a^0
    m_primitive_roots_log[0] = -1; //There is no exponent to make a^n == 0!

    for (size_t i = 1; i < m_element_count; ++i) {
        alpha <<= 1; //times x
        m_primitive_roots[i] = (Z2Polynomial(alpha) % irreducible.value()).coefficients();
        m_primitive_roots_log[m_primitive_roots[i].coefficients()] = i;
    }
}

Z2Polynomial FiniteField::multiply_roots(const Z2Polynomial& multiplier, const Z2Polynomial& multiplicand) const {
    if (multiplier.is_zero() || multiplicand.is_zero())
        return {};
    assert(multiplier.degree() < m_exponent && multiplicand.degree() < m_exponent);

    const auto alpha_exponent = m_primitive_roots_log[multiplier.coefficients()] + m_primitive_roots_log[multiplicand.coefficients()];
    return m_primitive_roots[alpha_exponent % m_element_count];
}

Z2Polynomial FiniteField::power_of_root(const Z2Polynomial& root, u8 exponent) const {
    assert(root.degree() < m_exponent);
    const auto alpha_exponent = m_primitive_roots_log[root.coefficients()];
    return m_primitive_roots[(alpha_exponent * exponent) % m_element_count];
}

size_t FiniteField::root_exponent(const Z2Polynomial& root) const {
    assert(root.degree() < m_exponent);
    return m_primitive_roots_log[root.coefficients()];
}

Z2Polynomial FiniteField::power_of_x(u8 exponent) const {
    return m_primitive_roots[exponent % m_element_count];
}

Z2Polynomial FiniteField::root_inverse(const Z2Polynomial& root) const {
    assert(!root.is_zero()); //There is no multiplicative inverse to 0

    const auto alpha_exponent = m_primitive_roots_log[root.coefficients()];
    return m_primitive_roots[(m_element_count - alpha_exponent) % m_element_count];
}

Z2Polynomial FiniteField::syndrome(const Z2Polynomial& polynomial, u8 n) const {
    Z2Polynomial result;
    for (auto exponent : polynomial.exponent_values())
        result += power_of_x(exponent * n);

    return result;
}
